sphinx/ext/extlinks.py

```diff
@@ -50,18 +50,25 @@ def role(typ: str, rawtext: str, text: str, lineno: int,
                 % (typ, base_url), line=lineno)
             full_url = base_url + part
         if not has_explicit_title:
-            if prefix is None:
+            if caption is None:
                 title = full_url
             else:
-                title = prefix + part
+                try:
+                    title = caption % part
+                except (TypeError, ValueError):
+                    inliner.reporter.warning(
```

Comment:
Could you use `sphinx.util.logging` module instead?

Reply:
Sure, I'll then also update the other ``inliner.reporter.warning`` usages in this extension.

I looked at `sphinx.util.logging` but it seems that in order to produce a location i need a filename and a line number.
Sadly, the ``role()`` function only gets passed the line number but no filename.  How do I gain access to this information?

Sidenote: Even better would be getting the filename and line number of the config file where the corresponding `extlinks` config-entry is "declared" (last assigned(?)) during the setup-phase but that seems to be impossible as the config is a python script...

Reply:
You need to pass a tuple of the current docname and lineno to the logging module. And you can get it via `inliner.document.settings.env.docname` call.

```
docname = inliner.document.settings.env.docname
logger.warning(__('...'), location=(docname, lineno))  # __(...) function is a helper for i18n
```

Reply:
It would be better to check the configuration on the generating role function. Emitting warnings on every extlink-role call is a bit noisy.

```
def make_link_role(base_url: str, caption: str) -> RoleFunction:
  # check the caption is a prefix (old format).
  if '%s' not in caption:
    # convert to the new one
    caption += '%s'

    logger.warning(...)  # emit a warning (if needed, please see another comment)

  def role(...):
     ...
```

In this case, no location info is not needed for the warning. It's an error of configuration, not source text's.

Reply:
Thanks!